### PR TITLE
Adding Julia Oxygen

### LINKED ISSUE #6736: [Julia] Add Oxygen
Oxygen is a newer framework for Julia. Would be interesting to see how fast is it.
Do you mind to add framework to this repo @ndortega?

https://docs.juliahub.com/Oxygen/JtS3f/1.1.2/tutorial/first_steps/
https://github.com/ndortega/Oxygen.jl

### PR DESCRIPTION
**Pull request by AbrJA**
Adding [Oxygen](https://github.com/OxygenFramework/Oxygen.jl) to the benchmark. I let the same structure as [Merly](https://github.com/neomatrixcode/Merly.jl)

------------

Closes #6736

### CODE REVIEW

**Comment by waghanza** [comment]
@ndortega Are you OK to introduce Oxygen on here ?

File: julia/oxygen/Project.toml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[deps]
+Oxygen = "df9a0d86-3283-4920-82dc-4555fc0d1d8b"
+HTTP = "cd3eb016-35fb-5094-929b-558a96fad6f3"
+[compat]
+Oxygen = "^1.4.0"
```
Comment (waghanza) [inline julia/oxygen/Project.toml:5]:
```suggestion
Oxygen = "^1.5.0"
```
or

```suggestion
Oxygen = "~1.5.0"
```


Not sure which one guarantees to use the 1.5.x version and is more standard in the Julia community

File: julia/oxygen/Manifest.toml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (waghanza) [inline julia/oxygen/Manifest.toml:1]: Is this file really needed ?

Disclaimer: it is present for merly, but not aware of Julia usages

File: julia/oxygen/Project.toml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+[deps]
+Oxygen = "df9a0d86-3283-4920-82dc-4555fc0d1d8b"
```
Comment (waghanza) [inline julia/oxygen/Project.toml:2]: Do we need to keep hashes in this file ?
I mean computing the right version to use won't be done at startup by Julia ?

File: julia/oxygen/server.jl [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+@get "/user/{id}" function(request::HTTP.Request, id)
+    id
+end
+
+@post "/user" function ()
+   ""
+end
+
+# Could be faster setting access_log = nothing and using multithreading with serveparallel
```
Comment (waghanza) [inline julia/oxygen/server.jl:19]: Maybe you could implement this, and also make it for merly.

I don't know about Julia usage, but multi threading seems to be, in some languages, done in specialized libraries, like gunicorn in python, or in explicit implementation like cluster.ts for bun in javascript

**Review by waghanza — CHANGES_REQUESTED** [review]

**Comment by ndortega** [comment]
@waghanza  Thanks for the ping!

I just opened up a similar PR on the techempower benchmarks repo: https://github.com/TechEmpower/FrameworkBenchmarks/pull/8789

The benchmark here is missing a couple of important performance tweaks: 
- setting `access_log=nothing` will turn off internal logging by HTTP which is a huge speedup
- Explicit type definitions in request handlers
- Using render functions like `text`, `json` to format responses in the request handlers. 
- Launching julia with 2 threads and running it in parallel mode (multithreading). I've noticed that in benchmarking the julia scheduler does much better with 2 threads for small requests. 

I'd like to see changes done before getting merged so that Oxygen can be judged on it's performance at it's best. 

Other notes: 

- We shouldn't include the Manifest.toml (this locks in specific package versions) 
- We can drop the compact section altogether so that it always pulls the latest.

File: julia/oxygen/Project.toml (hunk 3 above)
Comment (ndortega) [reply julia/oxygen/Project.toml:2]: We do need the UUID's - In Julia this is used to find the right package in the general registry. It's not tied to a specific version, that's what the `compat` section is for. If the `compat` section is dropped, then it will pull the latest version automatically

**Comment by waghanza** [comment]
Thanks for the clarification @ndortega and practices for the Julia community 

Let do it @AbrJA 

For information, the difference between tech empower and this project is that the first is more developer oriented (letting developers tweak the framework to have more performance). 

Here the approach is different, we take care of consistency, code readiness, version update (framework and languages), compilation options ... to reflect a production usage

File: julia/oxygen/Manifest.toml (hunk 2 above)
Comment (ndortega) [reply julia/oxygen/Manifest.toml:1]: Yup, we shouldn't need this file

File: julia/oxygen/Project.toml (hunk 1 above)
Comment (ndortega) [reply julia/oxygen/Project.toml:5]: I'd recommend dropping the compat section or at least changing it to `Oxygen = "^1"` so it grabs the latest 1.x.x version

**Comment by waghanza** [comment]
However, about 

> Launching julia with 2 threads and running it in parallel mode (multithreading). I've noticed that in benchmarking the julia scheduler does much better with 2 threads for small requests.

The idea here is to reflect what Julia could do in production. Not sure that using two threads is good (nor the opposite afaik), maybe using the number of CPU cores is accurate. Just to define a value that you are confident to use in a production environment, not a value driven by the idea of performance in a benchmark environment.

File: julia/oxygen/config.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+framework:
+  github: OxygenFramework/Oxygen.jl
+  version: 1.5.1
```
[the remark this answers is not in the document]
Comment (ndortega) [reply julia/oxygen/config.yaml:3]: Is the `version` property required?  If possible, I'd like to keep things as generic as possible so that it always pulls the latest version. Granted, I don't know if this is best practice in the benchmarking world but It would certainly make my life easier since I wouldn't have to make as many followup PR's for new versions

**Comment by ndortega** [comment]
> However, about
> 
> > Launching julia with 2 threads and running it in parallel mode (multithreading). I've noticed that in benchmarking the julia scheduler does much better with 2 threads for small requests.
> 
> The idea here is to reflect what Julia could do in production. Not sure that using two threads is good (nor the opposite afaik), maybe using the number of CPU cores is accurate. Just to define a value that you are confident to use in a production environment, not a value driven by the idea of performance in a benchmark environment.

In that case running it with `julia -t auto` will let julia decide how many threads to use based on the current hardware. 

But for running in production I'd recommend tweaking this to an actual value based on your workload. If your API has quick operations (under 50ms expected response time) fewer threads will give you better performance due to less overhead in the scheduler. On the other end, if you're expecting to handle heavy traffic with longer durations then have multiple threads would benefit you more.

**Comment by waghanza** [comment]
I think auto will be enough for now, at least because other languages are configured by doing so, for example javascript

File: julia/oxygen/Project.toml (hunk 3 above)
Comment (AbrJA) [reply julia/oxygen/Project.toml:2]: Done! [5fae3de](https://github.com/the-benchmarker/web-frameworks/pull/7286/commits/5fae3de3237836c57e57c9822b1b1e3b59d46993)

File: julia/oxygen/Manifest.toml (hunk 2 above)
Comment (AbrJA) [reply julia/oxygen/Manifest.toml]: [5fae3de](https://github.com/the-benchmarker/web-frameworks/pull/7286/commits/5fae3de3237836c57e57c9822b1b1e3b59d46993)

**Comment by waghanza** [comment]
Sorry not to have figured it out previously, maybe, we should keep d'EPS section in Toml file, and remove deps.

At least for me, it appears clearer to understand the version using X.Y.Z instead of a checksum.

Also I think that http module in Julia is a dependency of oxygen no ?

@ndortega @AbrJA

File: julia/oxygen/Project.toml [hunk 6]
```diff
@@ -0,0 +1,3 @@
...
+HTTP = "cd3eb016-35fb-5094-929b-558a96fad6f3"
```
Comment (waghanza) [inline julia/oxygen/Project.toml:3]: Maybe this should be dropped. This is already defined in Oxygen's project.toml

File: julia/oxygen/Project.toml (hunk 1 above)
Comment (AbrJA) [reply julia/oxygen/Project.toml]: [19cb81c](https://github.com/the-benchmarker/web-frameworks/pull/7286/commits/19cb81c4ae75855de69891e5069c9ed36b8d9a5f)

File: julia/oxygen/Project.toml (hunk 6 above)
Comment (waghanza) [reply julia/oxygen/Project.toml:3]: @AbrJA

Comment (AbrJA) [reply julia/oxygen/Project.toml:3]: I tried to avoid it but I got an error, let me check another way, I will test it locally to be sure

Comment (AbrJA) [reply julia/oxygen/Project.toml:3]: I was reviewing the implementation of @ndortega and he also include this into the [Project.toml](https://github.com/TechEmpower/FrameworkBenchmarks/pull/8789/files#diff-ad1f5944da489952393483f4aee90304851db60a10fa8dfa9263bfde68677495) Maybe the best option is to keep it

File: julia/oxygen/server.jl (hunk 4 above)
Comment (AbrJA) [reply julia/oxygen/server.jl]: [78dde22](https://github.com/the-benchmarker/web-frameworks/pull/7286/commits/78dde2266f45d603d35dc862d3f6fd95cd0ab974)

File: julia/oxygen/Project.toml (hunk 6 above)
Comment (waghanza) [reply julia/oxygen/Project.toml:3]: seems redundant for me, but let @ndortega gives his opinion

**Comment by cyrusmsk** [comment]
Thanks @AbrJA for implementing it. I was curious how will it work some time ago https://github.com/OxygenFramework/Oxygen.jl/issues/131

File: julia/oxygen/config.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (waghanza) [reply julia/oxygen/config.yaml:3]: This is just a descriptive version (to have this info displayed on results) and not the version installed

File: julia/oxygen/Project.toml (hunk 6 above)
Comment (AbrJA) [reply julia/oxygen/Project.toml:3]: @ndortega What do you think?

**Review by waghanza — APPROVED** [review]
